### PR TITLE
Added optional PoW method

### PR DESCRIPTION
**Pull request by IoTMOD**

### CODE REVIEW

File: pkg/pow/pow.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+const powsrvReinitCooldown = 30 * time.Second
+
+// Handler struct
+type Handler struct {
+	sync.RWMutex
+
+	powsrvClient      *powsrvio.PowClient
+	localPoWFunc      pow.ProofOfWorkFunc
+	powType           string
+	lastInitTimestamp int64
```
Comment (muXxer) [inline pkg/pow/pow.go]: Use time.Time

File: pkg/pow/pow.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+// tryReinitPowsrvWithoutLocking tries to reinit powsrv.io if the connection got lost.
+// the lock needs to be handled by the caller
+func (h *Handler) tryReinitPowsrvWithoutLocking() {
+	if h.powsrvClient == nil {
+		return
+	}
+
+	if h.lastInitTimestamp+powsrvReinitCooldown.Nanoseconds() <= time.Now().UnixNano() {
```
Comment (muXxer) [inline pkg/pow/pow.go]: Use time.Since

File: pkg/pow/pow.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+		h.powsrvClient.Close()
+		h.powsrvClient.Init()
+		h.lastInitTimestamp = time.Now().UnixNano()
+	}
+}
+
+// NewHandler creates a new PoW handler instance
+func NewHandler() (*Handler, error) {
```
Comment (muXxer) [inline pkg/pow/pow.go]: Call it "New"

File: pkg/pow/pow.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+	var localPoWFunc pow.ProofOfWorkFunc
```
Comment (muXxer) [inline pkg/pow/pow.go]: Create a pow plugin and init the Handler globally with:

```go
var (
	PLUGIN      = node.NewPlugin("PoW", node.Enabled, configure, run)
	log         *logger.Logger
	handler     *pow.Handler
	handlerOnce sync.Once
)


// Handler gets the pow Handler instance.
func Handler() *pow.Handler {
	handlerOnce.Do(func() {
		// init the pow handler with all possible settings
	})
	return handler
}
```

Then you can call `pow.Handler()` from all plugins that use it, and you can connect the shutdown event cleanly with its own priority.

File: pkg/pow/pow.go (hunk 3 above)
Comment (muXxer) [inline pkg/pow/pow.go]: Why return an error if it can never happen?

File: pkg/pow/pow.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+// GetPoWType returns the fastest available PoW type which gets used for PoW requests
+func (h *Handler) GetPoWType() string {
+	return h.powType
+}
+
+// DoPoW calculates the PoW
+// Either with the fastest available local PoW function or with the help of powsrv.io (optional, POWSRV_API_KEY env var must be available)
+func (h *Handler) DoPoW(trytes trinary.Trytes, mwm int, parallelism ...int) (nonce string, err error) {
+	h.Lock()
```
Comment (muXxer) [inline pkg/pow/pow.go]: Only one PoW request at a time would be possible with this.
Do we need a lock here?

File: plugins/coordinator/plugin.go [hunk 6]
```diff
@@ -68,15 +69,20 @@ func configure(plugin *node.Plugin) {
 	tangleplugin.SetUpdateSyncedAtStartup(true)
 
 	var err error
-	coo, err = initCoordinator(*bootstrap, *startIndex)
+	powHandler, err = pow.NewHandler()
```
Comment (muXxer) [inline plugins/coordinator/plugin.go]: Use the proposed global PoW handler from the plugin.

File: plugins/spammer/plugin.go [hunk 7]
```diff
@@ -68,6 +70,13 @@ func configure(plugin *node.Plugin) {
 	checkPeersConnected = node.IsSkipped(coordinator.PLUGIN)
 	spammerAvgHeap = utils.NewTimeHeap()
 
+	var err error
+	if powHandler, err = pow.NewHandler(); err != nil {
```
Comment (muXxer) [inline plugins/spammer/plugin.go]: Use the proposed global PoW handler from the plugin.

File: plugins/spammer/plugin.go [hunk 8]
```diff
@@ -132,6 +141,14 @@ func run(_ *node.Plugin) {
 		return
 	}
 
+	// close the PoW handler on shutdown
+	daemon.BackgroundWorker("Close PoW handler", func(shutdownSignal <-chan struct{}) {
```
Comment (muXxer) [inline plugins/spammer/plugin.go]: This can be moved to the global PoW handler plugin.

File: plugins/webapi/pow.go [hunk 9]
```diff
@@ -53,6 +52,15 @@ func attachToTangle(i interface{}, c *gin.Context, _ <-chan struct{}) {
 		return
 	}
 
+	// Init the PoW handler
+	powHandler, err := pow.NewHandler()
```
Comment (muXxer) [inline plugins/webapi/pow.go]: Use the proposed global PoW handler from the plugin.

**Review by muXxer — CHANGES_REQUESTED** [review]

File: pkg/whiteflag/test/whiteflag_test.go [hunk 10]
```diff
@@ -34,6 +34,7 @@ import (
 	"github.com/gohornet/hornet/pkg/profile"
 	"github.com/gohornet/hornet/pkg/utils"
 	"github.com/gohornet/hornet/pkg/whiteflag"
+	hornet_pow "github.com/gohornet/hornet/plugins/pow"
```
Comment (alexsporn) [inline pkg/whiteflag/test/whiteflag_test.go]: plugin in a package? 🙈

Comment (muXxer) [reply pkg/whiteflag/test/whiteflag_test.go]: Fixed :carlos:

**Review by alexsporn — APPROVED** [review]

**Review by muXxer — APPROVED** [review]

**Pull request merged**